pkg/agent/addon_status_controller.go

```diff
@@ -0,0 +1,141 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	addonv1alpha1 "open-cluster-management.io/api/addon/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const (
+	operatorDegradedReason    = "NoOperatorAvailable"
+	externalDNSDegradedReason = "NoExternalDNSAvailable"
+)
+
+// AddonStatusController reconciles Hypershift addon status
+type AddonStatusController struct {
+	spokeClient client.Client
+	hubClient   client.Client
+	log         logr.Logger
+	addonNsn    types.NamespacedName
+}
+
+// AddonStatusPredicateFunctions defines which Deployment this controller should watch
+var AddonStatusPredicateFunctions = predicate.Funcs{
+	CreateFunc: func(e event.CreateEvent) bool {
+		deployment := e.Object.(*appsv1.Deployment)
+		return containsHypershiftAddonDeployment(*deployment)
+	},
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		oldDeployment := e.ObjectOld.(*appsv1.Deployment)
+		newDeployment := e.ObjectNew.(*appsv1.Deployment)
+		return containsHypershiftAddonDeployment(*oldDeployment) && containsHypershiftAddonDeployment(*newDeployment)
+	},
+	DeleteFunc: func(e event.DeleteEvent) bool {
+		deployment := e.Object.(*appsv1.Deployment)
+		return containsHypershiftAddonDeployment(*deployment)
+	},
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (c *AddonStatusController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.Deployment{}).
+		WithEventFilter(AddonStatusPredicateFunctions).
+		Complete(c)
+}
+
+// Reconcile updates the Hypershift addon status based on the Deployment status.
+func (c *AddonStatusController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	c.log.Info(fmt.Sprintf("reconciling Deployment %s", req))
+	defer c.log.Info(fmt.Sprintf("done reconcile Deployment %s", req))
+
+	deployment := &appsv1.Deployment{}
+	err := c.spokeClient.Get(ctx, req.NamespacedName, deployment)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			c.log.Error(err, "failed to get the deployment")
+			return ctrl.Result{}, err
+		}
+
+		deployment.Name = req.Name
+	}
+
+	// generate addon status condition based on the Deployment
+	addonCondition := checkDeployment(*deployment, err)
```

Comment:
So if there are both operator and external-dns deployments, will the addon status reflect only one of them?

Reply:
Thanks Roke. I think I know what you are trying to get at. Let me improve this part so for every deployment it can better reflect both status.

Reply:
@rokej I fixed the code based on your comment. It can now show both not available if we are checking for external dns.

```
  - lastTransitionTime: "2022-11-01T02:40:30Z"
    message: |-
      There are no operator replica available
      There are no external-dns replica available
    reason: NoOperatorAvailable,NoExternalDNSAvailable
    status: "True"
    type: Degraded
```

and if  one of them comes back online then in the old code it will set degraded to false which is incorrect. With new code it will still shows that its degraded for the one that is not available.

```
  - lastTransitionTime: "2022-11-01T02:40:30Z"
    message: There are no operator replica available
    reason: NoOperatorAvailable
    status: "True"
    type: Degraded
```

Reply:
The `checkDeployments(()` code is a bit procedure and repetitive but it's easy to understand so I left it that way. It's also abstracted out to a helper function with no client connection requirement so I can use the function test code generator to do a good test coverage on it.